common/autoware_universe_utils/src/geometry/geometry.cpp

```diff
@@ -390,4 +390,425 @@ bool intersects_convex(const Polygon2d & convex_polygon1, const Polygon2d & conv
+{
+  const auto & vertices = poly.vertices();
+  const auto num_of_vertices = vertices.size();
+  int64_t winding_number = 0;
+
+  const auto [y_min_vertex, y_max_vertex] = std::minmax_element(
+    vertices.begin(), vertices.end(), [](const auto & a, const auto & b) { return a.y() < b.y(); });
+  if (point.y() < y_min_vertex->y() || point.y() > y_max_vertex->y()) {
+    return false;
+  }
+
+  double cross;
+  for (size_t i = 0; i < num_of_vertices; ++i) {
+    const auto & p1 = vertices[i];
+    const auto & p2 = vertices[(i + 1) % num_of_vertices];
+
+    if (p1.y() <= point.y() && p2.y() >= point.y()) {  // upward edge
+      cross = (p2 - p1).cross(point - p1);
+      if (cross > 0) {  // point is to the left of edge
+        winding_number++;
+        continue;
+      }
+    } else if (p1.y() >= point.y() && p2.y() <= point.y()) {  // downward edge
+      cross = (p2 - p1).cross(point - p1);
+      if (cross < 0) {  // point is to the left of edge
+        winding_number--;
+        continue;
+      }
+    } else {
+      continue;
+    }
+
+    if (std::abs(cross) <= std::numeric_limits<double>::epsilon()) {  // point is on edge
+      return true;
+    }
+  }
+
+  return winding_number != 0;
+}
+
+bool disjoint(const alt::ConvexPolygon2d & poly1, const alt::ConvexPolygon2d & poly2)
+{
+  if (equals(poly1, poly2)) {
+    return false;
+  }
+
+  if (intersects(poly1, poly2)) {
+    return false;
+  }
+
+  for (const auto & vertex : poly1.vertices()) {
+    if (touches(vertex, poly2)) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+double distance(
+  const alt::Point2d & point, const alt::Point2d & seg_start, const alt::Point2d & seg_end)
+{
+  const auto seg_vec = seg_end - seg_start;
+  const auto point_vec = point - seg_start;
+
+  const double seg_vec_norm = seg_vec.norm();
+  const double seg_point_dot = seg_vec.dot(point_vec);
+
+  if (seg_vec_norm <= std::numeric_limits<double>::epsilon() || seg_point_dot < 0) {
+    return point_vec.norm();
+  } else if (seg_point_dot > std::pow(seg_vec_norm, 2)) {
+    return (point - seg_end).norm();
+  } else {
+    return std::abs(seg_vec.cross(point_vec)) / seg_vec_norm;
+  }
+}
+
+double distance(const alt::Point2d & point, const alt::ConvexPolygon2d & poly)
+{
+  if (covered_by(point, poly)) {
+    return 0.0;
+  }
+
+  // TODO(mitukou1109): Use plane sweep method to improve performance?
+  const auto & vertices = poly.vertices();
+  double min_distance = std::numeric_limits<double>::max();
+  for (size_t i = 0; i < vertices.size(); ++i) {
+    min_distance =
+      std::min(min_distance, distance(point, vertices[i], vertices[(i + 1) % vertices.size()]));
+  }
+
+  return min_distance;
+}
+
+bool equals(const alt::Point2d & point1, const alt::Point2d & point2)
+{
+  return std::abs(point1.x() - point2.x()) <= std::numeric_limits<double>::epsilon() &&
+         std::abs(point1.y() - point2.y()) <= std::numeric_limits<double>::epsilon();
```

Comment:
From what I read (e.g., [here](https://en.cppreference.com/w/cpp/types/numeric_limits/epsilon) or [here](https://stackoverflow.com/questions/48133572/what-can-stdnumeric-limitsdoubleepsilon-be-used-for)), using `std::numeric_limits<double>::epsilon()` directly is not a good way to check for precision when comparing two numbers.
I think it would make more sense to define an epsilon based on the accuracy we want to reach (e.g., `1e-3` for mm accuracy).